### PR TITLE
Pass structs by readonly reference for Kind and IsKind

### PR DESCRIPTION
**Pull request by sharwell**
These methods are part of the public API, so we can't simply remove the old method. However, the benefits of pass-by-reference for extension methods are not possible if we simply add the second overload. This pull request changes the original public API to no longer be an extension method (a source breaking change) but still has the same signature (not a binary breaking change). A new extension method is added, addressing the source breaking change by causing the compiler to bind to the new method instead of the old.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Are there perf benefits here?  i'm def curious abou this because it feels like it would be unbelievably great  if the runtime could figue this out already.  i.e. syntax SyntaxToken is readonly, and presumably it can see that these mehods don't mutate, that it could do all this without copies.  but i get that that might be too much and we need the actual APIs.

But i am curious what sort of change we see (even if it's in a microbenchmark).

**Comment by sharwell** [comment]
I'm working on a much larger benchmark where changes like this are piped through the entire code style layer. These two methods are showing up in the analyzers as triggering hidden copies in hundreds of locations, but they are also public APIs. The change proposed here is intended to determine if the approach is viable with respect to our backwards compatibility policy, and if so, I would proceed with the larger-scale testing.

**Comment by CyrusNajmabadi** [comment]
Great.  Thanks for the explanation!

File: src/Compilers/CSharp/Portable/CSharpExtensions.cs [hunk 1]
```diff
@@ -14,12 +14,20 @@ namespace Microsoft.CodeAnalysis
 {
     public static class CSharpExtensions
     {
-        public static bool IsKind(this SyntaxToken token, SyntaxKind kind)
+        [Obsolete("Pass by reference instead.")]
+        public static bool IsKind(SyntaxToken token, SyntaxKind kind)
+            => token.IsKind(kind);
+
+        public static bool IsKind(this in SyntaxToken token, SyntaxKind kind)
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/CSharpExtensions.cs]: Is it possible to call this overload in reduced form? (i.e. `ref var token; token.IsKind();`)

Comment (sharwell) [reply src/Compilers/CSharp/Portable/CSharpExtensions.cs]: I was not able to find a way to call it without eliminating the other overload.

**Comment by sharwell** [comment]
@jcouv Removing **PR For Personal Review Only**. I need input from the compiler team in order to proceed.

**Comment by RikkiGibson** [comment]
@sharwell Do you still need feedback on this? If so please consider marking the PR as ready for review.

**Comment by sharwell** [comment]
@RikkiGibson I got some feedback on it. Now I have some additional work to determine the impact it will have in production scenarios.

**Pull request closed without merging**